File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.java
import java.util.Stack;

class Solution {
    public boolean parseBoolExpr(String expression) {
        Stack<Character> stack = new Stack<>();
        for (char c : expression.toCharArray()) {
            if (c == ')') {
                Stack<Boolean> exp = new Stack<>();
                while (stack.peek() != '(') exp.push(stack.pop() == 't');
                stack.pop();  // Pop the '('
                char op = stack.pop();  // Get the operator before '('
                boolean res = exp.pop();
                if (op == '&') while (!exp.isEmpty()) res &= exp.pop();
                else if (op == '|') while (!exp.isEmpty()) res |= exp.pop();
                else res = !res;
                stack.push(res ? 't' : 'f');
            } else if (c != ',') {
                stack.push(c);
            }
        }
        return stack.pop() == 't';
    }
}

File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        stack = []
        for c in expression:
            if c == ')':
                exp = []
                while stack[-1] != '(':
                    exp.append(stack.pop() == 't')
                stack.pop()  # Pop the '('
                op = stack.pop()  # Get the operator before '('
                res = exp.pop()
                if op == '&':
                    while exp: res &= exp.pop()
                elif op == '|':
                    while exp: res |= exp.pop()
                else:
                    res = not res
                stack.append('t' if res else 'f')
            elif c != ',':
                stack.append(c)
        return stack.pop() == 't'
